File: app/src/main/java/com/example/ExceptPains/Utils/PairTask.java
package com.example.ExceptPains.Utils;

import android.util.Log;

import java.util.HashMap;
import java.util.Map;
import java.util.Random;
import java.util.WeakHashMap;

public class PairTask {
    private static Map<Long, Noticeable> eventMap = new HashMap<>();
    private static Random randomGenerator = new Random();

    public interface Noticeable {
        public void onReceiveTaskResult(long eid, Object extra);
    }

    // 注册并返回一个id
    public static long observe(Noticeable identity) {
        Log.d("Utils.PairTask.observe", "entered.");

        long eid = -1;
        synchronized (eventMap) {
            while (eid < 0 || eventMap.containsKey(eid)) {
                eid = randomGenerator.nextLong();
            }
            eventMap.put(eid, identity);
        }

        Log.d("Utils.PairTask.observe", "before returning (" + eid + ")");
        return eid;
    }

    // 完成本次任务委托
    public static void finish(long eid, Object extra) {
        Log.d("Utils.PairTask.finished", "entered. eid is " + eid);

        Noticeable target;
        synchronized (eventMap) {
            target = eventMap.get(eid);
            eventMap.remove(eid);
        }
        if (target != null) {
            Log.d("Utils.PairTask.finished", "calling target.onReceive(" + eid + ", extra).");
            target.onReceiveTaskResult(eid, extra);
        }
    }
}

File: app/src/main/java/com/example/epledger/qaction/PairTask.kt
package com.example.epledger.qaction

import android.util.Log
import android.util.SparseArray
import java.util.*

object PairTask {
//    private val eventMap: MutableMap<Long, Noticeable?> = TreeMap()
    private val eventMap = SparseArray<Noticeable>(4)
    private val randomGenerator = Random()

    // 注册并返回一个id
    fun observe(identity: Noticeable?): Int {
//        Log.d("Utils.PairTask.observe", "entered.")
        var eid: Int = -1
        synchronized(eventMap) {
            while (eid < 0 || eventMap.get(eid) != null) {
                eid = randomGenerator.nextInt()
            }
            eventMap.put(eid, identity)
        }
//        Log.d("Utils.PairTask.observe", "before returning ($eid)")
        return eid
    }

    // 完成本次任务委托
    fun finish(eid: Int, extra: Any?) {
//        Log.d("Utils.PairTask.finished", "entered. eid is $eid")
        var target: Noticeable?
        synchronized(eventMap) {
            target = eventMap[eid]
            eventMap.remove(eid)
        }
        if (target != null) {
//            Log.d("Utils.PairTask.finished", "calling target.onReceive($eid, extra).")
            target!!.onReceiveTaskResult(eid, extra)
        }
    }

    interface Noticeable {
        fun onReceiveTaskResult(eid: Int, extra: Any?)
    }
}